### PR TITLE
MDEV-30454 Add CODING_STANDARDS.md file

### PR DESCRIPTION
**Pull request by LinuxJedi**
This is intended to be the start of a (not complete) coding standards document we can refer contributors to. This can be modified to add more nuances and become stricter over time. It can also have additional content for other file types (CMake, YACC, etc).

It does not cover plugins which should each individually have their own coding standards.

## Basing the PR against the correct MariaDB version
- [x] *This is a new feature and the PR is based against the latest MariaDB development branch*

### CODE REVIEW

**Comment by ericherman** [comment]
Hi @LinuxJedi,

Good to see some of this get documented.

I'd expect this to be referenced in `CONTRIBUTING.md` perhaps as a list item in "Help debug and develop MariaDB".

**Comment by LinuxJedi** [comment]
> Hi @LinuxJedi,
> 
> Good to see some of this get documented.
> 
> I'd expect this to be referenced in `CONTRIBUTING.md` perhaps as a list item in "Help debug and develop MariaDB".

I thought about that but figured it should be after this new file has been expanded by others. But I can do it in this PR instead.

**Comment by dlenski** [comment]
Great to see this!

Could you also update the Pull Request template (https://github.com/MariaDB/server/blob/HEAD/.github/pull_request_template.md) to reference these coding standards?

File: CODING_STANDARDS.md
Comment (dlenski) [inline CODING_STANDARDS.md:86]: It would be useful to recommend cleaning up whitespace in source files:

1. No trailing whitespace at end-of-line
2. No trailing blank lines at end-of-file
3. `\n` (POSIX) as end-of-line, not `\r\n` (Windows), for all text files
4. ***Some consistent standard for tabs vs. spaces for indentation*** so that all developers have a consistent view of the indentation of source files

I'm biased, but I'd also put in a plug for [dlenski/wtf](https://github.com/dlenski/wtf) as a tool that can do all these cleanups automatically, including [automatically doing them as part of a Git pre-commit hook](https://github.com/dlenski/wtf#git-pre-commit-hooks).

Comment (LinuxJedi) [reply CODING_STANDARDS.md:86]: I'll add these, but I don't want to add too much in this PR, otherwise it will never get merged. I want this to be more of a stub for other developers to add more things to.

Eventually I'd like to have a pre-commit hook and Buildbot test, it has not been decided which yet but I'll add that to the list. But I'd like narrower rules defined before we go that far, as well as some rules for some of the plugins. It will also likely require a clean-up pass of what we have. I'd also like to define as much as possible in a .editorconfig file too (mostly line endings and indentation) as many IDEs support this (and others can with a plugin), but plugins would need rule sets first.

This first step is intended to make things easier for new community members so that I have something to refer to. I literally wanted to refer contributors to this document a few times in the last few weeks to help them.

**Comment by LinuxJedi** [comment]
@ericherman cross links as suggested.

**Review by dlenski — DISMISSED** [review]

File: CODING_STANDARDS.md
Comment (dr-m) [inline CODING_STANDARDS.md:154]: `long` and `unsigned long` are not portable. `ssize_t` and `size_t` should be preferred. While most 64-bit ABIs that we target are LP64, Microsoft Windows follows LLP64, where `sizeof(long) == 4` and `sizeof(size_t) == 8`.
I think that we should explicitly discourage the use of `long` and `ulong`.

https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models

File: CODING_STANDARDS.md
Comment (dlenski) [inline CODING_STANDARDS.md:87]: Missing word: "should *be defined"

Comment (ericherman) [reply CODING_STANDARDS.md:87]: good catch

File: CODING_STANDARDS.md
Comment (LinuxJedi) [reply CODING_STANDARDS.md:154]: @dr-m Agreed, good call! In fact "char" even has portability issues as it can be signed or unsigned depending on platform :)

Comment (ericherman) [reply CODING_STANDARDS.md:154]: @dr-m @LinuxJedi this highlights something that should be its own issue: documentation lacks real guidance on when to use which types. Depending on the usage `size_t` can even be a very poor choice, for example in serialization one might choose `uint64_t` (or perhaps even `uint_least64_t`) in order to exchange data more easily between 32bit and 64bit platforms.

Regardless, rather than expand the scope of this change, I'd suggest that highlighting that preferred types are defined in `include/my_global.h`.

Comment (ericherman) [reply CODING_STANDARDS.md:154]: > @dr-m Agreed, good call! In fact "char" even has portability issues as it can be signed or unsigned depending on platform :)

For certain, and in real life others have been hit by that one, e.g.:
https://github.com/skarnet/skalibs/commit/52b7288713ecb189fdd473f9dfc7f9a8c4fb1ec2

Comment (vaintroub) [reply CODING_STANDARDS.md:154]: I think that  ulong is not good to use in any new API would be good enough. It is good for exactly nothing, neither serialization (fixed types are good), nor buffer sizes (size_t is good), nor file lengths or offsets (64bit intergers are good). ulong either brings warnings, and casts to silence the warnings, or it can be replaced with uint.

Comment (dlenski) [reply CODING_STANDARDS.md:154]: Why not use the [stdint.h types](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/stdint.h.html) when a specific size is needed?

- `uint32_t`, `int8_t` when you need a specific size (e.g. for serialization)
- `uintptr_t`, `intptr_t` when you need an integer of a size that can be round-trip casted/type-punned with a `void *`

Lots of FLOSS C/C++ projects make extensive use of these nowadays.

Comment (LinuxJedi) [reply CODING_STANDARDS.md]: > Why not use the [stdint.h types](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/stdint.h.html) when a specific size is needed?
> 
>     * `uint32_t`, `int8_t` when you need a specific size (e.g. for serialization)
> 
>     * `uintptr_t`, `intptr_t` when you need an integer of a size that can be round-trip casted/type-punned with a `void *`
> 
> 
> Lots of FLOSS C/C++ projects make extensive use of these nowadays.

There used to be portability issues with this a long time ago. I don't think it is so much of an issue any more for POSIX and Windows at least (definitely still a problem on other platforms that we don't really support). It could be revisited, but way out of scope for this PR. This is more of s descriptive PR of what is happening right now. Debates on changing the source code's style can happen later :)

File: CODING_STANDARDS.md
Comment (dr-m) [inline CODING_STANDARDS.md:33]: What does the Allman style say about the `switch` statement? In our code base, the opening brace for `switch` is a special case:
```c
switch (x) {
case X:
```

File: CODING_STANDARDS.md
Comment (dr-m) [inline CODING_STANDARDS.md:154]: I think that `size_t` and `ptrdiff_t` deserve to be mentioned explicitly. Their size is the integer register width, that is, 32 or 64 bits.

File: CODING_STANDARDS.md
Comment (LinuxJedi) [reply CODING_STANDARDS.md:33]: Hmm... yea, Allman isn't that. We also have a special case for 'if' as well by the look of Nikita's work I'll need to add.

**Comment by ericherman** [comment]
@LinuxJedi suggestion: separating sentences with newlines within paragraphs.

In my day-job, we have a lot of docs in markdown.
One-sentence-per-line still flows the same when rendered as separating sentences by period and a space.
Having each sentence on a separate line makes looking at `diff` output easier.
I hope that there will be many contributions to this file over time, thus it is worth making review easier.

**Comment by LinuxJedi** [comment]
> @LinuxJedi suggestion: separating sentences with newlines within paragraphs.
> 
> In my day-job, we have a lot of docs in markdown. One-sentence-per-line still flows the same when rendered as separating sentences by period and a space in regular paragraphs (bulleted lists are different, of course) Having each sentence on a separate line makes looking at `diff` output easier. I hope that there will be many contributions to this file over time, thus it is worth making review easier.

Good call, implemented :)

**Comment by haidong** [comment]
I suggest we incorporate @cvicentiu 's comments on [git practices and coding style](https://github.com/MariaDB/server/pull/2372#pullrequestreview-1261220304) into the `CODING_STANDARDS.md` file.

**Comment by LinuxJedi** [comment]
> I suggest we incorporate @cvicentiu 's comments on [git practices and coding style](https://github.com/MariaDB/server/pull/2372#pullrequestreview-1261220304) into the `CODING_STANDARDS.md` file.

Good all, I've added some key parts in :)

File: CODING_STANDARDS.md
Comment (ericherman) [inline CODING_STANDARDS.md:11]: grammar: "long standard git standard"

Perhaps intended to say "long standing git standard"?

Also I would say "de facto standard" or replace standard with "convention".

File: CODING_STANDARDS.md
Comment (ericherman) [inline CODING_STANDARDS.md:228]: combine this with "assignment" above

**Review by ericherman — CHANGES_REQUESTED** [review]
Overall, this is really good.

I found one blocking grammar change "long standard git standard" which I think should be fixed.

I also found some minor nit-picks, which should not block merging.


I would recommend that if the grammar error is fixed, that we merge this, then iterate to make it better.

File: CODING_STANDARDS.md
Comment (FooBarrior) [reply CODING_STANDARDS.md:228]: I think that `(const char*) db_name;` should be written without extra space after closing parenthesis, since type convertion is an unary operator

File: CODING_STANDARDS.md
Comment (LinuxJedi) [reply CODING_STANDARDS.md:11]: I think I meant "long standing git standard". But I'll rephrase.

File: CODING_STANDARDS.md
Comment (LinuxJedi) [reply CODING_STANDARDS.md:228]: > I think that `(const char*) db_name;` should be written without extra space after closing parenthesis, since type convertion is an unary operator

Since our usage here is mixed I'll remove that part of the specification for now until it can be decided as to what is correct.

**Review by ericherman — APPROVED** [review]

**Review by grooverdan — DISMISSED** [review]

**Comment by LinuxJedi** [comment]
Rebased off 11.0 (which unfortunately undid all the reviews)

**Comment by ottok** [comment]
This has now 7 commits: https://patch-diff.githubusercontent.com/raw/MariaDB/server/pull/2272.patch

Could part of the coding standard be that when review is requested, the git commit/patch to be reviewed is as final as possible and not spread across multiple ones in vain? :)

**Comment by LinuxJedi** [comment]
Squashed the commits

**Comment by LinuxJedi** [comment]
> Could part of the coding standard be that when review is requested, the git commit/patch to be reviewed is as final as possible and not spread across multiple ones in vain? :)

Maybe, but let's get the base in before we start adding more things.

**Comment by ericherman** [comment]
> when review is requested, the git commit/patch to be reviewed is as final as possible

This would be a great addition, but is not a coding style, rather more for the `pull_request_template.md` or perhaps `CONTIRBUTING.md`.

Yet, I agree with the idea; there is room for documenting more of these kinds of contributor expectations  -- perhaps the current content of the new `CODING_STANDARDS.md` should become a `## Code style` section of the `CONTRIBUTING.md` document.

Regardless, I support @LinuxJedi comment that it would be good to see maintainers get this current scope in; we can iterate to expand the scope.

Thus, if we decide it would be better to have a single file with larger scope, a follow-up PR can do that.

File: CODING_STANDARDS.md
Comment (ericherman) [inline CODING_STANDARDS.md:225]: @LinuxJedi Is the space after the cast valid?

I thought valid:
```c
name= (const char*)db_name;
name= (const char *)db_name;
```
Not valid:
```c
name= (const char*) db_name;
name= (const char *) db_name;
```

Am I just wrong?

**Review by ericherman — APPROVED** [review]

File: CODING_STANDARDS.md
Comment (LinuxJedi) [reply CODING_STANDARDS.md:225]: @ericherman the main part of the codebase uses every variation of spacing, sometimes in a single file, we don't appear to have a rule there right now.

This would likely fall under things we _should_ create a rule for but one does not exist yet. That stuff will definitely come later :)

Comment (ericherman) [reply CODING_STANDARDS.md:225]: Okay. Great. And yes, let's not invent new rules as part of this. Maybe that can be a separate follow up, but let's not scope-creep this with "we should" stuff.

**Pull request merged**
